### PR TITLE
jsonbuilder: avoids one allocation in encode_string

### PR DESCRIPTION
**Pull request by catenacyber**
Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
None specific, part of https://redmine.openinfosecfoundation.org/issues/4125

Describe changes:
- Improves performance of jsonbuilder `encode_string` by avoiding an allocation

Profiling was done on running fuzz_sigpcap_aware on its public corpus.
On my machine, `encode_string` takes about 17 seconds (I removed the inlining to get this)
With this patch, it takes about 2 seconds

### CODE REVIEW

File: rust/src/jsonbuilder.rs [hunk 1]
```diff
@@ -166,7 +166,7 @@ impl JsonBuilder {
             State::None => {
                 debug_validate_fail!("invalid state");
                 Err(JsonError::InvalidState)
-            },
```
Comment (catenacyber) [inline rust/src/jsonbuilder.rs:169]: rustfmt did this

File: rust/src/jsonbuilder.rs [hunk 2]
```diff
@@ -503,61 +503,34 @@ impl JsonBuilder {
...
-                offset += 1;
-                buf[offset] = b'0';
-                offset += 1;
-                buf[offset] = b'0';
-                offset += 1;
-                buf[offset] = HEX[(x >> 4 & 0xf) as usize];
-                offset += 1;
-                buf[offset] = HEX[(x & 0xf) as usize];
-                offset += 1;
+                let mut buf = vec![b'\\', b'u', b'0', b'0', 0, 0];
```
Comment (jasonish) [inline rust/src/jsonbuilder.rs:517]: This could be `let but = &mut &[b'\\', ...]` for maybe even tinier bit more performance..

**Comment by jasonish** [comment]
Nice to see this is faster.. Its also more what this should look like anyways.

File: rust/src/jsonbuilder.rs (hunk 2 above)
Comment (catenacyber) [reply rust/src/jsonbuilder.rs:517]: Then I get the compile error :
```
error[E0594]: cannot assign to `buf[_]` which is behind a `&` reference
   --> src/jsonbuilder.rs:518:17
    |
518 |                 buf[4] = HEX[(x >> 4 & 0xf) as usize];
    |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot assign
```

**Comment by jasonish** [comment]
Just to some up some rather manual bench marking.  This is much faster for small values, but for values over 1000 characters what is in master appears to be faster, especially as more and more bytes need to be encoded.  For some configurations that might be faster such as alerts only, no payload logging, etc.  But as soon as you added payload logging to your alerts master will perform better.

File: rust/src/jsonbuilder.rs (hunk 2 above)
Comment (jasonish) [reply rust/src/jsonbuilder.rs:517]: Oops, that should be:
```
let buf = &mut [b'\\', ...];
```

**Comment by catenacyber** [comment]
Replaced by https://github.com/OISF/suricata/pull/6196

File: rust/src/jsonbuilder.rs (hunk 2 above)
Comment (catenacyber) [reply rust/src/jsonbuilder.rs:517]: Thanks Jason

**Pull request closed without merging**